src/EdgeCollapsingConn.cpp

```diff
@@ -5,9 +5,146 @@
+        m_tet_connectivity[t_id].m_is_removed = true;
+        vector_erase(m_vertex_connectivity[v2_id].m_conn_tets, t_id);
+        vector_erase(n1_t_ids, t_id); // to add to conn_tets for v2
+        //
+        for (int j = 0; j < 4; j++) {
+            int v_id = m_tet_connectivity[t_id][j];
+            if (v_id != v1_id && v_id != v2_id)
+                vector_erase(m_vertex_connectivity[v_id].m_conn_tets, t_id);
+        }
+    }
+    //
+    for (size_t t_id : n1_t_ids) {
+        m_vertex_connectivity[v2_id].m_conn_tets.push_back(t_id);
+        int j = m_tet_connectivity[t_id].find(v1_id);
+        assert(j >= 0);
+        m_tet_connectivity[t_id][j] = v2_id;
+    }
+    vector_sort(m_vertex_connectivity[v2_id].m_conn_tets);
+    //
+    m_vertex_connectivity[v1_id].m_conn_tets.clear(); // release mem
+
+    Tuple new_loc = tuple_from_vertex(v2_id);
+    // todo: update timestamp of tets and tuple, change locs to loc
+    if (!collapse_after(new_loc)) {
+        m_vertex_connectivity[v1_id].m_is_removed = false;
+        for (int t_id : n12_t_ids) m_tet_connectivity[t_id].m_is_removed = false;
+        //
+        for (int i = 0; i < old_tets.size(); i++) {
+            int t_id = old_tets[i].first;
+            m_tet_connectivity[t_id] = old_tets[i].second;
+        }
+        for (int i = 0; i < old_vertices.size(); i++) {
+            int v_id = old_vertices[i].first;
+            m_vertex_connectivity[v_id] = old_vertices[i].second;
+        }
+
+        //        //fortest
+        //        std::cout<<old_m_vertex_connectivity.size()<<"
+        //        "<<m_vertex_connectivity.size()<<std::endl; for(int
+        //        i=0;i<old_m_vertex_connectivity.size();i++){
+        //            if(!(old_m_vertex_connectivity[i] == m_vertex_connectivity[i])){
+        //                std::cout<<i<<std::endl;
+        //                old_m_vertex_connectivity[i].print_info();
+        //                std::cout<<"//"<<std::endl;
+        //                m_vertex_connectivity[i].print_info();
+        //            }
+        //        }
+        //
+        //        std::cout<<old_m_tet_connectivity.size()<<"
+        //        "<<m_tet_connectivity.size()<<std::endl; for(int
+        //        i=0;i<old_m_tet_connectivity.size();i++){
+        //            if(!(old_m_tet_connectivity[i] == m_tet_connectivity[i])){
+        //                std::cout<<i<<std::endl;
+        //                old_m_tet_connectivity[i].print_info();
+        //                std::cout<<"//"<<std::endl;
+        //                m_tet_connectivity[i].print_info();
+        //            }
+        //        }
+        //
+        //        assert(old_m_vertex_connectivity == m_vertex_connectivity);
+        //        assert(old_m_tet_connectivity == m_tet_connectivity);
+        //        //fortest
+
+        return false;
+    }
+
+    /// call invariants on all entities
+    if (false) // if any invariant fails
```

Comment:
why `if(false)`

Reply:
invariant checks which probably the same as the link condition.